src/cylinder.jl

```diff
@@ -25,12 +25,21 @@ function Base.in(x::SVector{N}, s::Cylinder{N}) where {N}
     return sum(abs2,d - p*s.a) ≤ s.r^2
 end
 
-function normal(x::SVector{N}, s::Cylinder{N}) where {N}
+function surfpt_nearby(x::SVector{N}, s::Cylinder{N}) where {N}
     d = x - s.c
-    p = dot(d, s.a)
-    p > s.h2 && return s.a
-    p < -s.h2 && return -s.a
-    return normalize(d - p*s.a)
+    p = d ⋅ s.a
+    q = d - p*s.a
+    lp = abs(p)
+    lq = norm(q)
+    if abs(lp-s.h2) < abs(lq-s.r)
+        nout = sign(p)*s.a
```

Comment:
Actually, no, that would break the symmetry.   Leave it as-is (just returns a zero `nout` vector in that case).  In the subpixel-averaging code, check for `iszero(nout)` and just skip averaging in that case.